ncm-afsclt/src/main/perl/afsclt.pm

```diff
@@ -2,997 +2,314 @@
+
             # new entry
-            push(@{$ipaddrs{$cell}}, $1);
+            push ( @{ $ipaddrs{$cell} }, $1 );
         }
     }
-    close(FH);
-    foreach $cell (keys(%ipaddrs)) {
-        @{$ipaddrs{$cell}} = sort(@{$ipaddrs{$cell}});
-        $self->Debug("CellServDB $cell -> @{$ipaddrs{$cell}}");
+    $localcelldb_fh->close();
+    foreach $cell ( keys(%ipaddrs) ) {
+        @{ $ipaddrs{$cell} } = sort( @{ $ipaddrs{$cell} } );
+        $self->debug( 1, "CellServDB $cell -> @{$ipaddrs{$cell}}" );
     }
+
     # read known cells
     @todo = ();
-    my @current = `fs listcell -numeric 2>&1`;
+    my $proc = CAF::Process->new( [ "fs", "listcell", "-numeric" ], log => $self, keeps_state => 1 );
 
-    if($? >> 8 ) {
-      $self->info("cannot read current AFS cell info via fs:\n".join('',@current));
-      return 1;
+    if ( $? >> 8 ) {
+        $self->error( "Cannot read current AFS cell info via fs:\n" . $proc->output() );
+        return 1;
     }
-    foreach (@current) {
+
+    foreach ( split ( /\n/, $proc->output() ) ) {
         chomp($_);
         next unless /^Cell\s+(\S+)\s+on hosts\s+(\S.+)\.$/;
         $cell = $1;
-        @hosts = split(/\s+/, $2);
+        @hosts = split ( /\s+/, $2 );
         $seen{$cell} = 1;
-        $self->Debug("fs listcell: $cell -> @hosts");
-        next unless $ipaddrs{$cell} && @{$ipaddrs{$cell}};
+        next unless $ipaddrs{$cell} && @{ $ipaddrs{$cell} };
         @hosts = sort(@hosts);
-        if ("@hosts" ne "@{$ipaddrs{$cell}}") {
-            push(@todo, $cell);
-            $self->Debug("cell info for $cell changed from @hosts to @{$ipaddrs{$cell}}");
+        if ( "@hosts" ne "@{$ipaddrs{$cell}}" ) {
+            push ( @todo, $cell );
+            $self->debug( 1, "Cell info for $cell changed from @hosts to @{$ipaddrs{$cell}}" );
         }
     }
+
     # check new cells
-    foreach $cell (keys(%ipaddrs)) {
+    foreach $cell ( keys(%ipaddrs) ) {
         next if $seen{$cell};
-        next unless @{$ipaddrs{$cell}};
-        push(@todo, $cell);
-        $self->info("new cell $cell with @{$ipaddrs{$cell}}");
+        next unless @{ $ipaddrs{$cell} };
+        push ( @todo, $cell );
+        $self->info("New cell $cell with @{$ipaddrs{$cell}}");
     }
 
     if (@todo) {
-        $error = 0;
-        foreach $cell (@todo) {
-            my $out=`fs newcell $cell @{$ipaddrs{$cell}} 2>&1`;
-	    if($? >> 8) {
-	      $self->Debug("error while updating AFS cell info for $cell: $out");
-	      $error++;
-	    }
+        if ($NoAction) {
+            $self->info("Would need to configure new cells @todo");
+        }
+        else {
+            $error = 0;
+            foreach $cell (@todo) {
+                $proc = CAF::Process->new( [ "fs", "newcell", $cell, @{ $ipaddrs{$cell} } ], log => $self );
+                $proc->run();
+                if ( $? >> 8 ) {
+                    $self->error( "Error while updating AFS cell info for $cell: " . $proc->output() );
```

Comment:
will run the command a second time. there is also no need for the `NoAction` block, `CAF::Process` is safe to use